### PR TITLE
feat(grainfmt): Add format in place or output to a new file

### PR DESCRIPTION
**Pull request by marcusroberts**
Added support for

--format-in-place to overwrite the contents of the file with the formatted output
-o <filename>. output formatted results to a different file

Because the formatter can run on stdin without a source filename specified it will ignore --format-in-place if no input filename is given.

### CODE REVIEW

File: cli/bin/grain.js [hunk 1]
```diff
@@ -143,6 +143,7 @@ program
     "--verbose",
     "print critical information at various stages of compilation"
   )
+  .graincOption("--format-in-place", "reformat in place")
```
Comment (ospencer) [inline cli/bin/grain.js:146]: I think you can move the option to the `format` command so it only applies when that subcommand is active.

**Review by ospencer — COMMENTED** [review]
Any thoughts on `--write`? I've liked that name a lot when using Prettier!

File: cli/bin/grain.js [hunk 2]
```diff
@@ -211,6 +211,7 @@ program
 program
   .command("format [file]")
   .description("format a grain file")
+  .addOption(new ForwardOption("--in-place", "reformat in place"))
```
Comment (ospencer) [inline cli/bin/grain.js:214]:
```suggestion
  .addOption(new ForwardOption("--in-place", "format in place"))
```

**Review by ospencer — APPROVED** [review]

File: compiler/src/utils/config.re [hunk 3]
```diff
@@ -624,3 +624,6 @@ let get_implicit_opens = () => {
     ]);
   };
 };
+
+let format_in_place =
```
Comment (peblair) [inline compiler/src/utils/config.re:628]: @marcusroberts @ospencer Should we make an issue to split out the grainformat config items into a module that's not in the compiler? It feels odd to have it here

Comment (ospencer) [reply compiler/src/utils/config.re:628]: Somewhat related—I hadn't realized that the tools inherit all of the compiler config. We should probably also add to that issue (or make an issue for that too). It seems like they only need the stdlib path and include paths to be able to enter the typechecking phase, but I think they should just take those flags and ignore the rest.

Comment (peblair) [reply compiler/src/utils/config.re:628]: #917 created. Feel free to edit with any clarifications you think are necessary

File: cli/bin/grain.js [hunk 4]
```diff
@@ -211,6 +211,7 @@ program
 program
   .command("format [file]")
   .description("format a grain file")
+  .addOption(new ForwardOption("--in-place", "format in place"))
```
Comment (peblair) [inline cli/bin/grain.js:214]: Should be reconciled with the `.forwardOption` change in #916 (if that is merged before this PR).

File: compiler/src/utils/config.rei [hunk 5]
```diff
@@ -193,3 +193,7 @@ type implicit_opens =
   | Gc_mod;
 
 let get_implicit_opens: unit => list(implicit_opens);
+
+/** Formatter rewrites file in place */
+
+let format_in_place: ref(bool);
```
Comment (peblair) [inline compiler/src/utils/config.rei:199]: @ospencer So do we want this moved into `grainformat.re`? Or is that left to #917?

Comment (ospencer) [reply compiler/src/utils/config.rei:199]: Yeah, let's get that moved before this is merged.

Comment (peblair) [reply compiler/src/utils/config.rei:199]: @marcusroberts If you could move this flag out of `config.re` and into `grainformat.re` then that'd be :100:

Comment (marcusroberts) [reply compiler/src/utils/config.rei]: That's done

**Review by peblair — APPROVED** [review]
Looks good!

**Review by ospencer — COMMENTED** [review]
Looks great! I merged #916 so you can use `.forwardOption` now :)

**Pull request merged**